File: code/mapVis.js
function updateOpacity(value) {
    var opacity = value / 100;
    var overlays = document.getElementsByClassName('leaflet-image-layer');
    for (var i = 0; i < overlays.length; i++) {
        overlays[i].style.opacity = opacity;
    }
}

class overlayHandler{
    overlays = {};
    map = null;
    currentlyChecked= [];
    currentType = null;
    currentSat = null;
    
    constructor(map){
        this.map = map;
        this.setListener();
    }
    
    
    addOverlays(overlays){
        this.overlays = overlays;
    }
    
    addOverlay(first_year,last_year,country,clusters,bounds){
        this.overlays[country] = new overlay(first_year,last_year,country,clusters,bounds[0],bounds[1]);
    }
    
    updateMap(year){
        if(this.currentSat && this.currentType){
            for (const country of this.currentlyChecked){
                this.overlays[country].updateMap(this.map,this.currentType,this.currentSat,year)
            }
        }
    }
    
    removeFromMap(){
        for (const country of this.currentlyChecked){
            this.overlays[country].removeCurrentOverlay(this.map,this.currentType,this.currentSat,year)
        }
    }
    
    setListener() {
        this.setMapListener()
        this.setSliderListener()
        this.setCheckboxesListener()
        this.setRadioListener(document.getElementsByName("sat"),this.updateSat.bind(this))
        this.setRadioListener(document.getElementsByName("typeVis"),this.updateTypeVis.bind(this))
    }

    setMapListener(){
        this.map.on('click', (e)=>{
            L.popup()
            .setLatLng(e.latlng)
            .setContent(e.latlng.toString())
            .openOn(this.map);
        });
    }
    
    setSliderListener(){
        let handler = this
        $('#year').on('input', function() {
            if(handler.getType()!="kmeans")handler.updateMap($(this).val());
            let span = document.getElementById("yearValue");
            span.innerHTML = "";
            span.appendChild(document.createTextNode($(this).val()));
        });
    }
    
    setCheckboxesListener(){
        let handler = this
        var checkboxes = [...document.getElementsByClassName('leaflet-control-layers-selector')]
                                                    .filter(check => check.type == "checkbox");
        for (var i = 0; i < checkboxes.length; i++){
            checkboxes[i].addEventListener('change', function() {
                let innerText = this.nextSibling.textContent.trim();
                if (this.checked) {
                    handler.notifyCheck(innerText);
                    for (let color of ["red","purple"])
                        handler.addMarkerListener(color);
                }else{
                    handler.notifyUncheck(innerText);
                }
            });   
        }
    }
    
    setRadioListener(radioList, callback){
        for(const radio of radioList){
            radio.addEventListener('change',()=>{
                if (radio.checked){
                    callback(radio.value === "null" ? null :radio.value)
                }
            });
        }  
    }
    
    
    update(){
        this.updateMap($('#year').val())
    }
    
    addMarkerListener(color){
        let nameMarker = "awesome-marker-icon-"+color+" awesome-marker leaflet-zoom-animated leaflet-interactive"
        const markers = [...document.getElementsByClassName(nameMarker)].filter(m => !m.dataset.hasEvent)
        for (let marker of markers){
            marker.addEventListener("click",()=>{
                let cpt = 0;
                let idInterval = setInterval(()=>{
                    cpt++;
                    console.log("Tour : ", cpt)
                    let img = document.getElementById(getImgName(color));
                    if(img){
                        if (color == "purple")updateHist($('#year').val());
                        if (this.currentSat)updateImage(this.currentSat);
                        clearInterval(idInterval);
                    }
                    if(cpt>=100)clearInterval();
                },10)
            });
            marker.dataset.hasEvent = true
        }
    }
    
    notifyCheck(country){
        this.currentlyChecked.push(country);
        this.update();
        
    }
    
    notifyUncheck(country){
        let del = this.currentlyChecked.splice(this.currentlyChecked.indexOf(country),1);
        if (this.overlays[del])
            this.overlays[del].removeCurrentOverlay(this.map);
    }
    
    
    updateTypeVis(typeVis){
        this.currentType = typeVis;
        typeVis == null ? this.removeFromMap() : this.update();
    }
    
    getType(){
        return this.currentType
    }
    
    updateSat(sat){
        console.log(this,sat)
        this.currentSat = sat;
        console.log("Nouveau sat",this.currentSat , sat)
        sat==null?this.removeFromMap():this.update();
    }
    
    
}

class overlay{
    imageOverlays = {"ntl_intensity":{},"kmeans":{}};
    currentOverlay = null;
    
    constructor(first_year, last_year, country, clusters, top_map, bot_map) {
        let path ="";
        const sats = ["DMSP", "VIIRS"];
        for (let sat of sats) {
            this.imageOverlays["ntl_intensity"][sat] = {}
            path = "../analysis/" + country;
            for (let i = first_year; i < last_year+1; i++) {
                let n = "/ntl_intensity_" + i + "_" + sat + ".png";
                this.imageOverlays["ntl_intensity"][sat][i] = L.imageOverlay(path + '/ntl_intensity/' + sat + "/" + n,
                    [[bot_map[0], top_map[1]], [top_map[0], bot_map[1]]],
                    {opacity: 0.6}
                );  
            }
            this.imageOverlays["kmeans"][sat] = L.imageOverlay(
                path  + "/kmeans_analysis/"+ sat + "/" + clusters + "_" + first_year + "-" + last_year + "/cluster_img_"+clusters +".png",
                [[bot_map[0], top_map[1]], [top_map[0], bot_map[1]]],
                {opacity: 0.6}
            );
        }
    }
    
    updateMap(map,visType,currentSat,year) {
        this.removeCurrentOverlay(map);
        let nextOverlay =  this.imageOverlays[visType][currentSat];
        if (visType == "ntl_intensity") nextOverlay = nextOverlay[year];
        console.log("Incroyable",nextOverlay,this.imageOverlays,year);
        this.currentOverlay = nextOverlay ? nextOverlay.addTo(map) : null;
        updateOpacity(document.getElementById('opacityRange').value);
    }
    
    removeCurrentOverlay(map) {
        if (this.currentOverlay) map.removeLayer(this.currentOverlay);
    }
    
}

function getImgName(color){
    return color == "purple"?"hist_img":"kmeans_img"
}

function updateHist(year){
    var img = document.getElementById("hist_img");
    if (img){
        let path = img.src.split("/")
        let file = path[path.length-1].split("_")
        file[file.length-1] = year + ".png" 
        path[path.length-1] = file.join("_")
        img.src = path.join("/");
    }
}
function updateImage(sat) {
    if(sat == "null"){
        var closeButton = document.getElementsByClassName("leaflet-popup-close-button")[0];
        console.log(closeButton);
        if(closeButton){
            console.log("bruh");
            closeButton.click() ;
            return;
        }
    }
    var img = document.getElementById("hist_img");
    if (img){
        let path = img.src.split("/");
        let file = path[path.length-1].split("_");
        file[2] = sat;
        path[path.length-1] = file.join("_");
        img.src = path.join("/");
        return;
    }
    img = document.getElementById("kmeans_img");
    if (img){
        let path = img.src.split("/");
        path[path.length -3] = sat;
        img.src = path.join("/");
        document.getElementById("kmeans_sat").innerText = sat == "DMSP" ? "DMSP" : sat == "VIIRS" ? "VIIRS" : "none" ;
        console.log(document.getElementById("kmeans_sat") , sat == "DMSP" ? "DMSP" : sat == "VIIRS" ? "VIIRS" : "none" );
        return;
    }
}

File: code/map_vis.py
import numpy as np
import folium
import argparse
import rasterio
import os

from matplotlib.colors import Normalize
from matplotlib.cm import viridis
from utils import get_max_resized
from kmeans import main_kmeans
from plp import combined
from PIL import Image
import math



class NoSuchFileForThoseYears(Exception):
    def __init__(self, first_year, last_year):
        super().__init__(f"ERREUR aucun fichier pour ses années premiére : {first_year} et derniére : {last_year}")

class map_vis():
    def __init__(self,countries,force=[],first_year=2000, last_year=2020):
        self.force = force
        self.countries = countries

        min_accepted = 2000
        max_accepted = 2020
        if(min_accepted <= first_year <= last_year <= max_accepted):
            self.first_year= first_year
            self.last_year = last_year
        elif min_accepted <= last_year <= first_year <= max_accepted:
            print("Année passée à l'envers, on les inverse")
            self.last_year = first_year
            self.first_year = last_year
        else :
            raise(NoSuchFileForThoseYears(first_year,last_year))
        
    def __call__(self):

        m = folium.Map(location=[0, 0], 
                    tiles= folium.TileLayer(attr="Openstreetmap",show=False,name="Openstreetmap"),
                    zoom_start=3,
                    attr='OpenStreetMap',
                    world_copy_jump=True    
        )
        folium.TileLayer(tiles='https://{s}.tile.openstreetmap.fr/osmfr/{z}/{x}/{y}.png',
                        attr="OpenStreetMap France",
                        name='OpenStreetMap France'
        ).add_to(m)

        folium.TileLayer(
            tiles='https://server.arcgisonline.com/ArcGIS/rest/services/World_Imagery/MapServer/tile/{z}/{y}/{x}',
            name='ESRI World Imagery',
            attr='Tiles © Esri',
            show=False
        ).add_to(m)

        folium.TileLayer(
            tiles='https://{s}.tile.opentopomap.org/{z}/{x}/{y}.png',
            name='OpenTopoMap',
            attr='Map data © OpenStreetMap contributors, SRTM | Map style: © OpenTopoMap (CC-BY-SA)',
            show=False
        ).add_to(m)






        initJs= """const handler  = new overlayHandler("""+m.get_name()+""")"""
        for name in self.countries:
            country_obj = country(name,
                force=self.force,
                first_year=self.first_year,
                last_year=self.last_year)
            country_obj(m)
            initJs += country_obj.get_init_js()
        self.add_custom_script(m,initJs)
        folium.LayerControl().add_to(m)
                # Liste de différentes icônes et couleurs
        markers = [
            {'location': [45.5236, -122.6750], 'icon': 'info-sign', 'color': 'red', 'popup': 'Red Info Sign'},
            {'location': [45.5244, -122.6699], 'icon': 'cloud', 'color': 'blue', 'popup': 'Blue Cloud'},
            {'location': [45.5215, -122.6764], 'icon': 'heart', 'color': 'green', 'icon_color': 'white', 'popup': 'Green Heart'},
            {'location': [45.5222, -122.6655], 'icon': 'home', 'color': 'black', 'angle': 45, 'popup': 'Black Home'},
            {'location': [45.5250, -122.6780], 'icon': 'star', 'color': 'pink', 'popup': 'Pink Star'}
        ]

        # Ajouter les marqueurs à la carte
        for marker in markers:
            folium.Marker(
                location=marker['location'],
                popup=marker['popup'],
                icon=folium.Icon(
                    icon=marker['icon'],
                    color=marker.get('color', 'blue'),
                    icon_color=marker.get('icon_color', None),
                    angle=marker.get('angle', 0)
                )
            ).add_to(m)
        m.save(os.path.join("index.html"))
        print("Map sauvegardée !")

    def add_custom_script(self,m,jsCountriesInit):
        custom_js= """
        <style>
        .leaflet-image-layer {
            /* old android/safari*/
            image-rendering: -webkit-optimize-contrast;
            image-rendering: crisp-edges; /* safari */
            image-rendering: pixelated; /* chrome */
            image-rendering: -moz-crisp-edges; /* firefox */
            image-rendering: -o-crisp-edges; /* opera */
            -ms-interpolation-mode: nearest-neighbor; /* ie */
        }
        </style>
        <div style="position: fixed; bottom: 50px; left: 50px; width: 230px; z-index: 9999;">
            <label for="opacityRange">Opacité des calques
            <input type="range" id="opacityRange" min="0" max="100" value="60" step="1" 
                oninput="updateOpacity(this.value)"></label>
            <label for="year">Année : <span id="yearValue" name="yearValue">""" + str(math.ceil((self.first_year+self.last_year)/2)) + """</span>
            <br/>
            <input type="range" id="year" name="year" oninput="updateHist(this.value)" default=\""""+str(self.first_year)+"""" min=\""""+str(self.first_year)+"""" max=\""""+str(self.last_year)+"""" />
            </label>
            <br/>
            <label for="DMSP">DMSP
            <input type="radio" id="DMSP" name="sat" value="DMSP" onClick="updateImage(this.value)">
            </label>
            <label for="VIIRS">VIIRS
            <input type="radio" id="VIIRS" name="sat" value="VIIRS" onClick="updateImage(this.value)" >
            </label>
            <label for="noneSat">Aucun
            <input type="radio" id="noneSat" name="sat" value="null" onClick="updateImage(this.value)">
            </label>
            <br/>
            <label for="ntl_intensity">Intensité
            <input type="radio" id="ntl_intensity" name="typeVis" value="ntl_intensity">
            </label>
            <label for="kmeans">Kmeans
            <input type="radio" id="kmeans" name="typeVis" value="kmeans">
            </label>
            <label for="noneType">Masquer
            <input type="radio"  id="noneType" name="typeVis" value="null">
            </label>
        </div>
        <script src="./mapVis.js"></script>
        <script>
            function updateHist(year){
                var img = document.getElementById("hist_img");
                if (img){
                    let path = img.src.split("/")
                    let file = path[path.length-1].split("_")
                    file[file.length-1] = year + ".png" 
                    path[path.length-1] = file.join("_")
                    img.src = path.join("/");
                }
            }
            function updateImage(sat) {
            console.log(sat)
                if(sat == "null"){
                    var closeButton = document.getElementsByClassName("leaflet-popup-close-button")[0];
                    console.log(closeButton)
                    if(closeButton){
                    console.log("bruh")
                        closeButton.click() = false
                    }
                }
                var img = document.getElementById("hist_img");
                if (img){
                    let path = img.src.split("/")
                    let file = path[path.length-1].split("_")
                    file[2] = sat
                    path[path.length-1] = file.join("_")
                    img.src = path.join("/");
                }
                img = document.getElementById("kmeans_img");
                if (img){
                    let path = img.src.split("/")
                    path[path.length -3] = sat 
                    img.src = path.join("/");
                    document.getElementById("kmeans_sat").innerText = sat == "DMSP" ? "DMSP" : sat == "VIIRS" ? "VIIRS" : "none" ;
                    console.log(document.getElementById("kmeans_sat") , sat == "DMSP" ? "DMSP" : sat == "VIIRS" ? "VIIRS" : "none" )
                }
            }
            document.addEventListener("DOMContentLoaded", function() {
                """+jsCountriesInit+"""
            });
        </script>"""
        m.get_root().html.add_child(folium.Element(custom_js))

class country():
    def __init__(self,name,cluster=5,floor=0,force=[],first_year=2000, last_year=2020):
        raster_path = os.path.join("../data", name, "DMSP", "ntl_2000.tif")
        with rasterio.open(raster_path) as dataset:
            width = dataset.width
            height = dataset.height
            transform = dataset.transform
        #transformation de mercator 
        top_left_coords = rasterio.transform.xy(transform, 0, 0)
        bottom_right_coords = rasterio.transform.xy(transform, height - 1, width - 1)
        self.bounds = [[top_left_coords[1], top_left_coords[0]], [bottom_right_coords[1], bottom_right_coords[0]]]

        self.name = name    
        self.floor = floor
        self.force = force
        self.cluster = cluster
        self.last_year = last_year
        self.first_year  = first_year
        self.cluster_dir_name = f"{cluster}_{first_year}-{last_year}"
        
    def __call__(self,m):
        print("Ajout à la carte de",self.name)
        li_sat = ["DMSP","VIIRS"]
        self.pre_generate_overlays(li_sat)
        self.generate_grp().add_to(m)

    def pre_generate_overlays(self,li_sat):
        for sat in li_sat:
            self.pre_generate_kmeans(sat )
            self.pre_generate_ntl_intensities(sat )
        self.pre_generate_plp()

    def pre_generate_kmeans(self,sat ):
        base_path = os.path.join("../analysis",self.name,"kmeans_analysis",sat,
                            self.cluster_dir_name)
        if (not os.path.exists(os.path.join(base_path,f"cluster_img_{self.cluster}.png"))
            or not os.path.exists(os.path.join(base_path,f"clusters_{self.cluster}.png") 
            or ("kmeans" in self.force))):
            print("Kmeans non trouvé, création de l'image...")
            params = {
                'ntl_type': sat,
                'resize': True,
                'show': False,
                'clusters': [self.cluster]
            }
            main_kmeans(self.name, **params, first_year=self.first_year, last_year=self.last_year)
            print("Image créée, reprise de la suite")

    def pre_generate_plp(self):
        graphs = self.generate_graphs()
        if len(graphs) > 0:
            plp = combined(self.name,floor= self.floor , first_year=self.first_year, last_year=self.last_year)
            plp(show=False,graphs=graphs)

    def generate_graphs(self):  
        if ("plp" in self.force):
            return ["graph","histogramme_annuel"]  
        
        graphs = []
        base_path = os.path.join("../analysis",self.name,"lit_pixel_analysis","DMSP_et_VIIRS")   
        if (not os.path.exists(os.path.join(base_path,str(self.floor),"lit_pixel_combined.png"))
            or ("graph" in self.force)) :
            print("Graph non trouvé, lancement de sa création")
            graphs.append('graph')
        if  ("histogramme_annuel" in self.force) :
            graphs.append('histogramme_annuel')
        else : 
            if(self.is_missing_hist()):
              graphs.append('histogramme_annuel')  
        return graphs
    
    def is_missing_hist(self):
        for sat in ["DMSP","VIIRS"]:
            for year in (self.first_year,self.last_year):
                if not os.path.exists(os.path.join("../analysis",self.name,"lit_pixel_analysis",
                                                        "DMSP_et_VIIRS","histogramme",
                                                        f"histogramme_{self.name}_{sat}_{year}.png")):
                    print(f"histogrammes annuels non trouvés pour {self.name} avec {sat} en {year}, lancement de leur création")
                    return True
        
    def pre_generate_ntl_intensities(self,sat ):
        vmax = -1
        path = f'../analysis/{self.name}/ntl_intensity/{sat}'
        os.makedirs(path,exist_ok=True)
        for year in range(self.first_year,self.last_year+1):
            temp_path = path + f"/ntl_intensity_{year}_{sat}.png"
            if not os.path.exists(temp_path) or ("ntl_intensity" in self.force):
                print(f"{self.name} : Création de l'image des intensité nocturnes pour {sat} de {year}")
                if vmax == -1 :
                    if sat =="DMSP":
                        vmax= 64
                    else:
                        vmax= get_max_resized(self.name)
                self.generate_ntl_intensity(sat,year,temp_path,vmax)

    def generate_ntl_intensity(self,sat,year,path,vmax):
        img_path = os.path.join("../data",self.name,sat,f"ntl_{year}.tif")
        with rasterio.open(img_path) as img:
            raster_data = img.read(1)
        #raster_data = np.load(img_path)
        norm = Normalize(vmin=0, vmax=vmax)
        ntl_intensity_data = viridis(norm(raster_data))
        ntl_intensity_image = (ntl_intensity_data[:, :, :3] * 255).astype(np.uint8)
        ntl_intensity_pil = Image.fromarray(ntl_intensity_image)
        ntl_intensity_pil.save(path)
    
    def generate_grp(self):
        grp_country = folium.FeatureGroup(name=f"{self.name}",show=False)

        folium.Marker(
            [self.bounds[0][0], self.bounds[0][1]], 
            popup=folium.Popup( """
        <!DOCTYPE html>
        <html>
            <body>
                <p>Carte des clusters kmeans <span id="kmeans_sat">DMSP</span></p>
                <img id="kmeans_img" src="../analysis/"""+self.name+"""/kmeans_analysis/DMSP/"""+self.cluster_dir_name+"""/clusters_"""+str(self.cluster)+""".png"
                alt="Image" width="700" height="150">
            </body>
        </html>
        """, max_width=1400, lazy=True),
            icon=folium.Icon(color='red', icon='info-sign')
        ).add_to(grp_country)
        
        folium.Marker(
            [self.bounds[0][0], self.bounds[0][1] + ((self.bounds[1][1] - self.bounds[0][1]) / 2)], 
            icon=folium.Icon(color='blue', icon='info-sign'),
            popup=folium.Popup(
            """<!DOCTYPE html>
            <html>
                <body>
                    <p>Graphique</p>
                    <img src="../analysis/"""+self.name+"""/lit_pixel_analysis/DMSP_et_VIIRS/"""+str(self.floor)+"""/lit_pixel_combined.png"
                    alt="Image" width="598" height="457">
                </body>
            </html>
            """,max_width=1400,lazy=True)
        ).add_to(grp_country)

        folium.Marker(
            [self.bounds[0][0], self.bounds[1][1]], 
            icon=folium.Icon(color='purple', icon='info-sign'),
            popup=folium.Popup(
            """
                <div>
                    <p>Histogramme</p>
                    <img id="hist_img" src="../analysis/"""+self.name+"""/lit_pixel_analysis/DMSP_et_VIIRS/histogramme/histogramme_"""
                    +self.name+"""_DMSP_"""+str(math.ceil((self.first_year+self.last_year)/2))+""".png" alt="Image" width="852" height="535">
                </div>
            """,lazy=True,max_width=2000)
        ).add_to(grp_country)

        folium.Rectangle(
        bounds=[self.bounds[0], self.bounds[1]],
            color='black',
            weight = 2.5
        ).add_to(grp_country)
        return grp_country

    def get_init_js(self ):
        return """
        handler.addOverlay("""+str(self.first_year)+""","""+str(self.last_year)+""",'"""+self.name+"""',"""+str(self.cluster)+""","""+str(self.bounds)+""")"""

if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("-n", "--name",nargs='+', help="Nom de la ou des zones d'études")
    parser.add_argument("-a", "--all",action='store_true', help="Prend toute les zones d'étude")
    parser.add_argument("-f","--force",nargs="+",default=[],help="Force la génération des objets générables(ntl_intensity,graph,kmeans)")
    parser.add_argument("-y","--year",nargs="+",default=[2000,2020],type=int,help="Années de début et de fin")
    args = parser.parse_args() 
    f = []
    for i in args.force:
        if i in ["kmeans", "k"]:
            f.append("kmeans")
        if i in ["plp", "p", "allumées"]:
            f.append("plp")
        if i in ["ntl_intensity", "ntl_i", "ntl", "n", "intensity", "i"]:
            f.append("ntl_intensity")
        if i in ["all", "tous", "a"]:
            f.append("kmeans")
            f.append("plp")
            f.append("ntl_intensity")
    
    if args.all:
        temp = []
        with os.scandir("../data") as files:
            for file in files:
                if file.is_dir():
                    temp.append(file.name)
    else:
        temp = args.name
    main = map_vis(temp,f, first_year=args.year[0], last_year=args.year[1])
    main()